filament/src/main/java/net/fabricmc/filament/task/RemapUnpickDefinitionsTask.java

```diff
@@ -92,58 +126,189 @@ public void execute() {
+					final UnpickV3Writer writer = new UnpickV3Writer();
+					reader.accept(new UnpickV3Remapper(writer) {
+						@Override
+						protected String mapClassName(String className) {
+							return mappingTree.mapClassName(className.replace('.', '/'), fromM, toM).replace('/', '.');
+						}
 
-					classMappings.put(
-							classFromName,
-							Objects.requireNonNull(classDef.getName(toM), "Null to name: " + classFromName)
-					);
-
-					for (MappingTree.MethodMapping methodDef : classDef.getMethods()) {
-						methodMappings.put(
-								new MethodKey(
-										Objects.requireNonNull(classFromName, "Null dst name: " + classDef.getSrcName()),
-										Objects.requireNonNull(methodDef.getName(fromM), "Null dst name: " + methodDef.getSrcName()),
-										Objects.requireNonNull(methodDef.getDesc(fromM), "Null dst name: " + methodDef.getSrcName())
-								),
-								Objects.requireNonNull(methodDef.getName(toM), "Null to name: " + methodDef.getSrcName())
-						);
-					}
+						@Override
+						protected String mapFieldName(String className, String fieldName, String fieldDesc) {
+							final MappingTree.FieldMapping fieldMapping = mappingTree.getField(className.replace('.', '/'), fieldName, fieldDesc, fromM);
 
-					for (MappingTree.FieldMapping fieldDef : classDef.getFields()) {
-						fieldMappings.put(
-								new FieldKey(
-										Objects.requireNonNull(classFromName, "Null dst name: " + classDef.getSrcName()),
-										Objects.requireNonNull(fieldDef.getName(fromM), "Null dst name: " + fieldDef.getSrcName())
-								),
-								Objects.requireNonNull(fieldDef.getName(toM), "Null to name: " + fieldDef.getSrcName())
-						);
-					}
-				}
+							if (fieldMapping == null) {
+								return fieldName;
+							}
+
+							final String dstName = fieldMapping.getName(toM);
+							return dstName == null ? fieldName : dstName;
+						}
 
-				try (UnpickV2Reader reader = new UnpickV2Reader(new FileInputStream(getParameters().getInput().getAsFile().get()))) {
-					UnpickV2Writer writer = new UnpickV2Writer();
-					reader.accept(new UnpickV2Remapper(classMappings, methodMappings, fieldMappings, writer));
+						@Override
+						protected String mapMethodName(String className, String methodName, String methodDesc) {
+							final MappingTree.MethodMapping methodMapping = mappingTree.getMethod(className.replace('.', '/'), methodName, methodDesc, fromM);
+
+							if (methodMapping == null) {
+								return methodName;
+							}
+
+							final String dstName = methodMapping.getName(toM);
+							return dstName == null ? methodName : dstName;
+						}
+
+						@Override
+						protected List<String> getClassesInPackage(String pkg) {
+							return jarIndex.classesInPackages.getOrDefault(pkg, List.of());
+						}
+
+						@Override
+						protected String getFieldDesc(String className, String fieldName) {
+							final String fieldKey = className + "." + fieldName;
+							String fieldDesc = jarIndex.fieldDescs.get(fieldKey);
```

Comment:
In loom I used TR for this, may that be easier? I'm slightly worried about how much code is duplicated (in a slightly different way) between here and loom. Long term the less code we have to maintain the better, but then I dont want to go too extreme with it.

https://github.com/FabricMC/fabric-loom/blob/573bc119ecf434666e0fbbfc1b2bbe9730905d29/src/main/java/net/fabricmc/loom/task/service/UnpickRemapperService.java#L137-L147

Reply:
Does a `TrClass` also include classes that are not in the mappings, even JDK classes?

Reply:
Not immediately sure how to reuse that. I'll leave it for now as this code will already have to change quite a bit when it becomes an Enigma plugin.

TR might be a bit easier, I didn't think of it, although I'm already having to scan the jar for the packages. Looking at it, it doesn't seem to include JDK classes, which I do via reflection here, it's required because Yarn has some unpick definitions referencing `java.util.Spliterator`.